### DISCUSSION TITLE
Performance and Hardware Configurations

### DISCUSSION THREAD

**Discussion by spectre-ns** [inventree/InvenTree#3166] [General]
Hey All,

I'm building out an inventree instance in AWS using a M5a.large instance. It's a little slow. I'm looking to get a feel of what other's are using to run their instances. Currently, I'm just rolling it out so it's just me working on it and I'm seeing slow loading times on database heavy pages. Looking forward to what every has to say!

**Comment by SchrodingersGat** [comment]
@spectre-ns what are you considering as "slow"?

- Running "bare metal" or docker?
- Which pages are you seeing as particularly slow?
- How many database entries do you have?
- Can you use the developer tools tab to determine which elements are taking a long time to load?

**Reply by spectre-ns** [reply]
10s to load a screen is slow in my opinion. Maybe my expectations are too high.

Running in docker.
Particularly the BOM page. This BOM is one level ~200 parts
Under 1000
See screen shot below. Looks server side because my system is sitting idle waiting for the results.
<img width="329" alt="image" src="https://user-images.githubusercontent.com/96780897/172949332-f9206496-f8ae-4f0c-8328-482463a57725.png">

**Reply by SchrodingersGat** [reply]
10s is an eternity!

Check if you have this option enabled:

![image](https://user-images.githubusercontent.com/10080325/172950191-99b85200-f2c8-4968-a942-94f1cfed1664.png)

There is currently an outstanding issue with price calculation, which is in desperate need of a refactor. If you have this option enabled, the BOM data can take a long time to serialize on the server

**Reply by spectre-ns** [reply]
@SchrodingersGat Thanks for your help on this. Still seems to be struggling... It struggles loading the part summary too.

<img width="298" alt="image" src="https://user-images.githubusercontent.com/96780897/172951757-efd3fa63-599f-4bfe-aa98-9e6d75f0347b.png">

**Reply by SchrodingersGat** [reply]
Is this still on the BOM page? Or a different one? Can you let me know the URL (minus your domain info) so I can compare?

**Reply by spectre-ns** [reply]
Same page... https://{domain}/part/30/# All parts of loading the part are slow.

**Reply by SchrodingersGat** [reply]
How does your loading time compare with the demo server https://demo.inventree.org

**Reply by spectre-ns** [reply]
Very similar it sits idle for a fair bit before responding loading the page. I guess this is the expected behavior. Sorry for all the questions just didn't seem as snappy as I would expect on a new basically empty system. It appears to be much faster on small assemblies.

<img width="242" alt="image" src="https://user-images.githubusercontent.com/96780897/172955627-d6d4b628-dd6f-430c-84eb-fdaa6a2d841d.png">

**Reply by spectre-ns** [reply]
I assume it loads the BOMs when you load the part and it's doesn't lazily load them only when you click the BOM option on the left hand side of the screen?

**Reply by matmair** [reply]
@spectre-ns you should be able to speed that up 10-100x with configuring redis - it is a DB bottleneck when no caching is enabled.

**Reply by SchrodingersGat** [reply]
@spectre-ns I am getting much shorter page loads on the demo server

https://demo.inventree.org/part/86/

![image](https://user-images.githubusercontent.com/10080325/172958168-46342c90-07ad-40e2-bb50-c67f42b36e46.png)

Your server response time of ~14s is well beyond what I have experienced (at least in the last 2 years after major refactoring :D)

Can you ascertain which request is taking 14s?

**Reply by spectre-ns** [reply]
Try this one please! https://demo.inventree.org/part/112/

🙂🙂🙂

**Reply by SchrodingersGat** [reply]
Hmm yeah that is glacial... It appears there are a few different things going on:

### Static Files

Static files seem to be taking an inordinate amount of time to retrieve

![image](https://user-images.githubusercontent.com/10080325/172965667-42351ea0-3335-4fe1-b09a-77059aefd606.png)

### BOM Data

The BOM query is also taking ages:

![image](https://user-images.githubusercontent.com/10080325/172965980-5aa78fe7-ce4d-4733-8843-607ab8beebf4.png)

Caching would help here, but this is something that requires investigation. It shouldn't be taking this long...

**Reply by spectre-ns** [reply]
I looked at top while making the requests and it looks like it's maxing out a single gunicorn worker for the whole request duration.

**Reply by SchrodingersGat** [reply]
I've made two new issues for this:

- https://github.com/inventree/InvenTree/issues/3168
- https://github.com/inventree/InvenTree/issues/3169

Any additional information you can provide here will help us track down these issues.
